Fixed #29746 -- Fixed misleading FlatpageForm URL help text if APPEND_SLASH is disabled.

knyghty opened the pull request:
Hijacking some existing tests since they're very tightly related, but let me know if these should have new tests.

Also not 100% sure on the implementation, but couldn't immediately see a better way.

knyghty commented on django/contrib/flatpages/forms.py:
```diff
@@ -22,6 +22,13 @@ class Meta:
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if settings.APPEND_SLASH and 'django.middleware.common.CommonMiddleware' in settings.MIDDLEWARE:
```
That's an interesting thought. I'll try it out and see what happens. I was originally concerned about how it would impact people changing settings at runtime (which isn't really supported but I know it happens), but I think the chance of people messing with middleware and `APPEND_SLASH` at runtime is pretty remote anyway.

knyghty commented:
Thanks Carlton. I don't really think the first test is necessary, as APPEND_SLASH does nothing without COMMON_MIDDLEWARE anyway, and is documented. I agree with your other points on reflection, and will get back with changes for those, but let me know if you think the other test is necessary.

carltongibson commented:
@knyghty I'm not particularly convinced myself that we need to cover that other case.

knyghty commented:
Hi @carltongibson 

I've addressed some of your concerns but I've hit a bit of a stumbling block trying to get the logic out of `__init__()`. I tried this:

```
url = forms.RegexField(
        label=_("URL"),
        max_length=100,
        regex=r'^[-\w/\.~]+$',
        help_text=(
            _help_text_with_trailing_slash
            if settings.APPEND_SLASH and
            'django.middleware.common.CommonMiddleware' in settings.MIDDLEWARE
            else _help_text_without_trailing_slash
        ),
        error_messages={
            "invalid": _(
                "This value must contain only letters, numbers, dots, "
                "underscores, dashes, slashes or tildes."
            ),
        },
    )
```

but unfortunately it makes my test fail:

```
======================================================================
FAIL: test_flatpage_doesnt_requires_trailing_slash_without_append_slash (flatpages_tests.test_forms.FlatpageAdminFormTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/Users/tom/work/personal/django/django/test/utils.py", line 371, in inner
    return func(*args, **kwargs)
  File "/Users/tom/work/personal/django/tests/flatpages_tests/test_forms.py", line 67, in test_flatpage_doesnt_requires_trailing_slash_without_append_slash
    "Example: '/about/contact'. Make sure to have a leading slash.",
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 829, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/unittest/case.py", line 822, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: "Example: '/about/contact/'. Make sure to have leading and trailing slashes." != "Example: '/about/contact'. Make sure to have a leading slash."

----------------------------------------------------------------------
```

I tried to reimport the form during the test case, thinking it's using the one imported when the settings were overridden differently, but this didn't seem to help anything.

Do you have any ideas here?

carltongibson commented:
Hmmm. (Initial thought: more work than justified by feature... 🙂) 

Yes: I guess the class is getting created with one settings, and then failing the test for the other. This would no doubt work in a real case, since it would only be one or the other. 

What's the best place to make it suitably dynamic? (May well be `__init__()`)

knyghty commented:
I think `__init__()` is okay, but I suppose we could also use a class method (or something else?) to test if we should have the slash.

This isn't really much better per se other than avoiding `___init__()` if it works, but it does mean we can reuse it in `clean_url()` rather than having the same check twice. What do you think?

knyghty commented:
Ah, never mind, it seems that will just end up with the same problem. I think the solution here is the best I can think of, annoyingly.

timgraham commented:
I think it's better not to move the existing help text (which might require retranslating it?).
